kiva/blend2d.py

```diff
@@ -0,0 +1,780 @@
+    # Color manipulation
+    # ----------------------------------------------------------------
+
+    def set_fill_color(self, color):
+        self.gc.set_fill_style(color)
+
+    def set_stroke_color(self, color):
+        self.gc.set_stroke_style(color)
+
+    def set_alpha(self, alpha):
+        self.gc.set_alpha(alpha)
+
+    # ----------------------------------------------------------------
+    # Gradients
+    # ----------------------------------------------------------------
+
+    def linear_gradient(self, x1, y1, x2, y2, stops, spread_method,
+                        units="userSpaceOnUse"):
+        """ Sets a linear gradient as the current brush.
+        """
+        gradient = blend2d.LinearGradient(x1, y1, x2, y2)
+        gradient.extend_mode = gradient_spread_modes.get(
+            spread_method, blend2d.ExtendMode.PAD
+        )
+        for stop in stops:
+            gradient.add_stop(stop[0], stop[1:])
+        self.gc.set_fill_style(gradient)
+
+    def radial_gradient(self, cx, cy, r, fx, fy, stops, spread_method,
+                        units="userSpaceOnUse"):
+        """ Sets a radial gradient as the current brush.
+        """
+        gradient = blend2d.RadialGradient(cx, cy, fx, fy, r)
+        gradient.extend_mode = gradient_spread_modes.get(
+            spread_method, blend2d.ExtendMode.PAD
+        )
+        for stop in stops:
+            gradient.add_stop(stop[0], stop[1:])
+        self.gc.set_fill_style(gradient)
+
+    # ----------------------------------------------------------------
+    # Drawing Images
+    # ----------------------------------------------------------------
+
+    def draw_image(self, img, rect=None):
+        """
+        img is either a N*M*3 or N*M*4 numpy array, or a PIL Image
+
+        rect - a tuple (x, y, w, h)
+        """
+        from PIL import Image
+
+        def normalize_image(img):
+            if not img.mode.startswith("RGB"):
+                img = img.convert("RGB")
+            return img
+
+        if isinstance(img, np.ndarray):
+            # Numeric array
+            img = Image.fromarray(img)
+            img = normalize_image(img)
+            img_array = np.array(img)
+        elif isinstance(img, Image.Image):
+            img = normalize_image(img)
+            img_array = np.array(img)
+        elif isinstance(img, GraphicsContext):
+            img_array = img._buffer
+        elif hasattr(img, "bmp_array"):
+            # An offscreen kiva.agg context
+            # XXX: Use a copy to kill the read-only flag which plays havoc
+            # with the Cython memoryviews used by blend2d
+            img = Image.fromarray(img.bmp_array)
+            img = normalize_image(img)
+            img_array = np.array(img)
+        else:
+            msg = "Cannot render image of type '{}' into blend2d context."
+            warnings.warn(msg.format(type(img)))
+            return
+
+        # XXX: Upside down!
+        dst_rect = blend2d.Rect(*rect)
```

Comment:
I don't think this will work in the default case where `rect=None` - I don't know if that default case is ever exercised in reality.

Reply:
You're right, that will raise an exception.